src/rust/engine/src/scheduler.rs

```diff
@@ -103,25 +103,54 @@ impl Scheduler {
   }
 
   ///
-  /// Attempts to complete a Node, recovering from `Failure::Invalidated` up to `count` times.
+  /// Attempts to complete all of the given roots, retrying the entire set (up to `count`
+  /// times) if any of them fail with `Failure::Invalidated`.
   ///
   /// In common usage, graph entries won't be repeatedly invalidated, but in a case where they
   /// were (say by an automated process changing files under pants), we'd want to eventually
   /// give up.
   ///
-  fn create(core: Arc<Core>, root: Root, count: usize) -> NodeFuture<Value> {
-    if count == 0 {
-      future::err(throw("Exhausted retries due to changed files.")).to_boxed()
-    } else {
-      core
-        .graph
-        .create(root.clone(), &core)
-        .or_else(move |e| match e {
-          Failure::Invalidated => Scheduler::create(core, root, count - 1),
-          x => future::err(x).to_boxed(),
+  fn execute_helper(
+    core: Arc<Core>,
+    roots: Vec<Root>,
+    count: usize,
+  ) -> BoxFuture<Vec<Result<Value, Failure>>, ()> {
+    // Attempt all roots in parallel, failing fast to retry for `Invalidated`.
+    let roots_res = future::join_all(
+      roots
+        .clone()
+        .into_iter()
+        .map(|root| {
+          core
+            .graph
+            .create(root.clone(), &core)
+            .then::<_, Result<Result<Value, Failure>, Failure>>(move |r| {
```

Comment:
Is the type ascription necessary?

Reply:
It was at some point, but I feel that it adds a bit of clarity in this case since we're double-resulting intentionally.